### PR TITLE
Added Xact verb

### LINKED ISSUE #116: implement xact_jun2010

### PR DESCRIPTION
**Pull request by ImperatorS79**
Should fix #116.

### CODE REVIEW

File: Engines/Wine/Verbs/xact/script.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        "dec2006_xact_x86.cab", "fev2006_xact_x86.cab", "fev2007_xact_x86.cab",
+        "fev2010_xact_x86.cab", "jun2006_xact_x86.cab", "jun2007_xact_x86.cab",
+        "jun2008_xact_x86.cab", "jun2010_xact_x86.cab", "mar2008_xact_x86.cab",
+        "mar2009_xact_x86.cab", "nov2007_xact_x86.cab", "nov2008_xact_x86.cab",
+        "oct2006_xact_x86.cab", 
+    ];
+    extractXacttoSystemDirectory(progressBar, filesToExtract, that.system32directory(), "xactengine*.dll", "xact_x86/");
+    
+    
+    //---------------------------------------------------------Extarct X3Daudio*.dll----------------------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        .extract(["-L", "-F", "*_x3daudio_*x86*"]);
+    
+    var filesToExtract = [
+        "feb2010_x3daudio_x86.cab", "jun2008_x3daudio_x86.cab", "mar2008_x3daudio_x86.cab",
+        "mar2009_x3daudio_x86.cab", "nov2007_x3daudio_x86.cab", "nov2008_x3daudio_x86.cab"
+    ];
+    extractXacttoSystemDirectory(progressBar, filesToExtract, that.system32directory(), "X3Daudio*.dll", "x3daudio_x86/");
+     
+     
+    //---------------------------------------------------------Extarct XAudio*.dll and XAPOFX*.dll---------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    regsvr32Xact(xactToRegserv);
+    regsvr32Xact(xaudioToRegserv);
+    
+    remove(this.prefixDirectory + "/drive_c/xact_x86/");
+    remove(this.prefixDirectory + "/drive_c/x3daudio_x86/");
+    remove(this.prefixDirectory + "/drive_c/xaudio_x86/");
+    
+    
+    if (this.architecture() == "amd64") {
+        //---------------------------------------------------------Extarct xactengine*.dll (x64)--------------------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            "dec2006_xact_x64.cab", "fev2006_xact_x64.cab", "fev2007_xact_x64.cab",
+            "fev2010_xact_x64.cab", "jun2006_xact_x64.cab", "jun2007_xact_x64.cab",
+            "jun2008_xact_x64.cab", "jun2010_xact_x64.cab", "mar2008_xact_x64.cab",
+            "mar2009_xact_x64.cab", "nov2007_xact_x64.cab", "nov2008_xact_x64.cab",
+            "oct2006_xact_x64.cab", 
+        ];
+        extractXacttoSystemDirectory(progressBar, filesToExtract, that.system64directory(), "xactengine*.dll", "xact_x64/");
+        
+        
+        //---------------------------------------------------------Extarct X3Daudio*.dll (x64)----------------------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            .extract(["-L", "-F", "*_x3daudio_*x64*"]);
+        
+        var filesToExtract = [
+            "feb2010_x3daudio_x64.cab", "jun2008_x3daudio_x64.cab", "mar2008_x3daudio_x64.cab",
+            "mar2009_x3daudio_x64.cab", "nov2007_x3daudio_x64.cab", "nov2008_x3daudio_x64.cab"
+        ];
+        extractXacttoSystemDirectory(progressBar, filesToExtract, that.system64directory(), "X3Daudio*.dll", "x3daudio_x64/");
+        
+        
+        //---------------------------------------------------------Extarct XAudio*.dll and XAPOFX*.dll (x64)---------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        .url("http://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe")
+        .checksum("f8f1217f666bf2f6863631a7d5e5fb3a8d1542df")
+        .name("directx_Jun2010_redist.exe")
+        .get();
+
+    var progressBar = this._wizard.progressBar(tr("Please wait ..."));
+    progressBar.setText(tr("Extracting {0} ...", "Xact"));
+    progressBar.setProgressPercentage(0.);
+
+    //---------------------------------------------------------Extarct xactengine*.dll--------------------------------------------
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: extract

File: Engines/Wine/Verbs/xact/script.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+* Inspired from d3dx9 verb, winetricks xact
+* -> https://github.com/Winetricks/winetricks/blob/63bc6dbe612d017a0cb6bf6e4cde265162d75bca/src/winetricks#L9643
+* @returns {Wine}
+*/
+Wine.prototype.xact = function () {
+    var that = this;
+
+    var extractXacttoSystemDirectory = function(progressBar, filesToExtract, destination, pattern, directory) {
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: The `filesToExtract` and `destination` are given as parameters. Therefore I think the function name is more complex than it has to be. I would prefer a more simple name like `extractFiles`. Also, please add some documentation for the function.

File: Engines/Wine/Verbs/xact/script.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            new CabExtract()
+                .archive(that.prefixDirectory + "/drive_c/"+ directory + cabFile)
+                .to(destination)
+                .extract(["-L", "-F", pattern]);
+
+            numberOfExtractedFiles++;
+        });
+    };
+    
+    var regsvr32Xact = function(dllToRegsvr) {
```
Comment (plata) [inline Engines/Wine/Verbs/xact/script.js]: Please add some documentation for the function.

**Comment by plata** [comment]
#116 requests xact_jun2010 (not just xact). I'm not exactly sure where so difference is so: is this solved by this PR?

File: Engines/Wine/Verbs/xact/script.js (hunk 7 above)
Comment (ImperatorS79) [reply Engines/Wine/Verbs/xact/script.js]: It's basically "extractDirectXtoSystemDirectory" from the directx9 verb

**Comment by ImperatorS79** [comment]
xact_jun2010 has been deprecated by winetricks, cf [https://github.com/Winetricks/winetricks/blob/63bc6dbe612d017a0cb6bf6e4cde265162d75bca/src/winetricks#L18895](https://github.com/Winetricks/winetricks/blob/63bc6dbe612d017a0cb6bf6e4cde265162d75bca/src/winetricks#L18895).

The implementation here is normaly the same as in winetricks

File: Engines/Wine/Verbs/xact/script.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+/**
+* Inspired from d3dx9 verb, winetricks xact
+* -> https://github.com/Winetricks/winetricks/blob/63bc6dbe612d017a0cb6bf6e4cde265162d75bca/src/winetricks#L9643
+* @returns {Wine}
+*/
+Wine.prototype.xact = function () {
+    var that = this;
+
+    var extractFiles = function(progressBar, filesToExtract, destination, pattern, directory) {
```
Comment (madoar) [inline Engines/Wine/Verbs/xact/script.js]: @plata I think I've seen such methods quite often, could we maybe extract it to remove duplication?

Comment (plata) [reply Engines/Wine/Verbs/xact/script.js]: Yes, why not. The question is where to put it. Either `Utils/Functions/Filesystem/Extract/` or a new one in `Utils/Functions/Filesystem`.

Comment (plata) [reply Engines/Wine/Verbs/xact/script.js]: Maybe `archive()` could receive a list?

Comment (madoar) [reply Engines/Wine/Verbs/xact/script.js]: I'm not really sure where to locate such a method, I would follow your advice for this.

Comment (plata) [reply Engines/Wine/Verbs/xact/script.js]: How about:
```Javascript
new BatchExtract()
    .extractor(new CabExtract())
    .archives(["archive1", "archive2"])
    .extract("arguments")
```

`BatchExtract` would be placed in the same file as `CabExtract`.

Comment (madoar) [reply Engines/Wine/Verbs/xact/script.js]: How do we pass the `progressBar`?

Comment (plata) [reply Engines/Wine/Verbs/xact/script.js]:
```Javascript
new BatchExtract()
    .extractor(new CabExtract())
    .archives(["archive1", "archive2"])
    .wizard(wizard)
    .extract("arguments")
```

Comment (plata) [reply Engines/Wine/Verbs/xact/script.js]: I merged this PR. Your proposal can be implemented separately. Feel free to provide a PR/open an issue for it.

**Pull request merged**